src/transformStateWithClones.js

```diff
@@ -7,7 +7,53 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stagesOfModify = [];
+  const modifyingState = { ...state };
+
+  for (const action of actions) {
+    const { type, ...rest } = action;
+    let modifiedObject = {};
+
+    switch (type) {
+      case 'clear':
+        modifiedObject = clear(modifyingState);
+        break;
+      case 'removeProperties':
+        modifiedObject = removeProperties(modifyingState, rest);
+        break;
+      case 'addProperties':
+        modifiedObject = addProperties(modifyingState, rest);
+        break;
+      default:
+        break;
+    }
+
+    stagesOfModify.push({ ...modifiedObject });
+  }
+
+  return stagesOfModify;
+};
+
+function clear(obj) {
+  for (const key of Object.keys(obj)) {
+    delete obj[key];
+  }
+
+  return obj;
+};
+
+function addProperties(obj, addedProperties) {
```

Comment:
In the addProperties function, consider renaming 'addedProperties' to 'extraData' for better clarity.